### PR TITLE
Feature/issue 760 cvodes refactor

### PR DESCRIPTION
**Pull request by wds15**
#### Submission Checklist

- [ ] Run unit tests: `./runTests.py test/unit`
- [X] Run cpplint: `make cpplint`
- [X] Declare copyright holder and open-source license: see below

#### Summary:
As the `coupled_ode_system` and the cvodes integration now work the same, we can merge the two systems. This pull removes much more code than it adds and adding varying initial times should be far easier now.

#### Intended Effect:
Only have a single ODE sub-system to generate the coupled ODE system.

#### How to Verify:

Run the involved tests:
```
test/unit//math/rev/mat/functor/integrate_ode_adams_prim_test.cpp
test/unit//math/rev/mat/functor/integrate_ode_adams_rev_test.cpp
test/unit//math/rev/mat/functor/integrate_ode_bdf_rev_test.cpp
test/unit//math/rev/mat/functor/integrate_ode_cvodes_grad_rev_test.cpp
test/unit//math/rev/mat/functor/integrate_ode_bdf_prim_test.cpp
test/unit//math/rev/mat/functor/cvodes_ode_data_rev_test.cpp
test/unit//math/rev/mat/functor/cvodes_ode_data_prim_test.cpp
```

#### Side Effects:

1. More speed! For a small 2cmt ODE (3 states, 5 parameters) I got ~20% speedup... I guess due to better memory locality of the `couple_ode_system` when comparing to the existing implementation.
2. The type of exception whenever an ODE system functor returns an inconsistent state has been aligned with the non-stiff system. That is, for a bad system functor which returns an inconsistent size, the `coupled_ode_system` has always thrown `invalid_argument` and we do now the same for the CVODES integrators
3. Note that the Jacobian for the stiff ODE integration just uses the `coupled_ode_system` for varying initials and known parameters... which is really just the Jacobian if evaluated at the state where we want it (a small comment is on the code).
4. `ode_system` has been removed such that user-supplied gradients for the ODE now must be done through the `coupled_ode_system` (but with the advantage now that the user only has to provide the `coupled_ode_system` with analytic gradients and he will speed up stiff and non-stiff integration). 

Additional notes for `cvodes_ode_data`:

The CVODES ode data serves as a callback object for CVODES. So CVODES gets thorugh this object three things:
1. ODE RHS, function  `rhs`
2. ODE sensitivity RHS, function `rhs_sens`
3. Jacobian wrt to states, function `dense_jacobian`

The old version did these things by calls to the `ode_system`.

The new version does this differently:
1. calling the user supplied functor with `double` versions of the parameters and the state
2. evaluating the respective `coupled_ode_system`. However, the `coupled_ode_system` pastes together the ODE RHS and the ODE sensitivity RHS. This is why the `rhs_sens` function skips the first `N` outputs (`N` being number of states or the length of the vector returned by the ODE RHS functor) and returns the rest of it.
3. The Jacobian wrt to the states *is* exactly equal to evaluation of the `coupled_ode_system<F,var,double>` (so varying initial states and known parameters) if we set the initial to be the state where we want to know the Jacobian. This done in line 131-140 of `cvodes_ode_data`.

The `cvodes_ode_data` has `prim` and `rev` tests defined which are the last two mentioned above.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Sebastian Weber

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by wds15** [comment]
The travis job failed due to a single job exceeding the time out limit while all others finished just fine.

@seantalts Can I trigger a restart of Travis (ideally only the one which had the time out)?

**Comment by wds15** [comment]
@seantalts sorry for the spam, just figured that I can log into Travis and restart myself.

**Comment by wds15** [comment]
grr... now Travis time outs again. Not sure what can be done about that. Looks like Travis is brittle.

**Comment by bob-carpenter** [comment]
Sorry about that.  Hopefully it'll eventually go through (and we'll eventually replace it).

**Comment by wds15** [comment]
I have now restarted some of the sub-jobs for about the 7. time or so. The time out issue is always coming up, I hope we can merge without Travis being green once everyone is happy with the pull.

**Comment by seantalts** [comment]
Travis is flaky but 7 in a row has been historically pretty unlikely... Do the tests take more time with this PR? I can split them up further if that's expected behavior.

**Comment by wds15** [comment]
I don't think that tests take longer than usual. Splitting sounds great. A number of test chunks come close to 50 minutes at which point it gets shot down which is happening quite a bit for me this time.

**Comment by seantalts** [comment]
I'm just asking because this test normally takes 37 minutes, so it's weird if it's taking more than 50 minutes 7 times in a row. Travis flakiness has normally been a one-off thing that is resolved on a retry, in my experience. So I am prone to worry that something now takes longer than it used to? Though these are the failing tests: `test/unit/math/prim/mat/err test/unit/math/prim/mat/meta test/unit/math/prim/mat/prob test/unit/math/prim/mat/vectorize`. I don't expect any of those are affected by this PR, right?

**Comment by wds15** [comment]
No, these have not been touched.

**Comment by seantalts** [comment]
Do you think the code those tests cover has changed?

**Comment by syclik** [comment]
@wds15, does this wiki page need to change? https://github.com/stan-dev/math/wiki/ODE-System-Refactor

**Comment by wds15** [comment]
Travis is finally green.

@seantalts not sure if I understand your question, but this pull does not touch those parts which were in question.

**Comment by wds15** [comment]
@syclik that wiki is fully out of date. In fact, in the past I opted to pull the coupled ode system over to use the ode_system thing. Now I am going the reverse as this seems to be the accepted ansatz to this problem.

In short: In the past the ode systems worked differently in that whenever initals vary this shifting thing was done in the non-stiff case (coupled ode system). We have removed this a few pulls ago which made the non-stiff system ~15% faster on one of my examples due to smaller AD stack size and even better this aligned the two systems (stiff / non-stiff).

You could help me if you could write specific questions to the "Side Effects" section bullets above. These points above were intended to explain what is going on. They were obviously too brief, so questions on this would help me where I need to clarify. Thanks.

**Comment by syclik** [comment]
Yes, will do. I'll have a bunch of questions for you later.

Thanks for doing this. I just need to wrap my head around it. I think the
first thing is the high-level code overview. Which things changed and why.

**Comment by bob-carpenter** [comment]
Please just remove wikis that are out of date.  Out of date is worse than nothing as it's misleading.  Thanks!

**Comment by wds15** [comment]
Ok, will get rid of that obsolete wiki.

The big change with this refactor is that I do not attempt to refactor both systems at the same time, but rather align CVODES with the non-stiff one. This hopefully makes it easier to finally get rid of the code duplication we have right now.

**Comment by wds15** [comment]
@syclik : feel free to comment on my "side effects" above before you waste time on this. If any of these points is nonsense to you, then I can try to rephrase. Also, whatever high-level doc is needed... let me know.

File: test/unit/math/rev/mat/functor/integrate_ode_adams_prim_test.cpp [hunk 1]
```diff
@@ -356,10 +356,9 @@ TEST(StanMathOde_integrate_ode_adams, error_conditions_bad_ode) {
   std::vector<int> x_int(2, 0);
 
   std::string error_msg
-      = "ode_system: size of state vector y (2) and derivative vector dy_dt "
-        "(3) "
-        "in the ODE functor do not match in size.";
+      = "cvodes_ode_data: dz_dt (3) and states (2) must match in size";
```
Comment (syclik) [inline test/unit/math/rev/mat/functor/integrate_ode_adams_prim_test.cpp]: I think the change in error type is good. I think we should revisit the error message. Is this clear? This will be pushed all the way to the user and I think the older message is clearer.

**Comment by syclik** [comment]
@wds15, the first level high level point is... what's being refactored? I think `stan/math/rev/mat/functor/cvodes_integrator.hpp`'s `integrate()` function keeps its interface and most of its behavior intact. Is that it?

**Comment by syclik** [comment]
Comments on side effects:
> - More speed! For a small 2cmt ODE (3 states, 5 parameters) I got ~20% speedup... I guess due to better memory locality of the couple_ode_system when comparing to the existing implementation.

Can you send an example? And which implementation are you talking about? What function (`integrate_ode_???`)?

> - The type of exception whenever an ODE system functor returns an inconsistent state has been aligned with the non-stiff system. That is, for a bad system functor which returns an inconsistent size, the coupled_ode_system has always thrown invalid_argument and we do now the same for the CVODES integrators

Great!

> - Note that the Jacobian for the stiff ODE integration just uses the coupled_ode_system for varying initials and known parameters... which is really just the Jacobian if evaluated at the state where we want it (a small comment is on the code).

Where is this?

> - ode_system has been removed such that user-supplied gradients for the ODE now must be done through the coupled_ode_system (but with the advantage now that the user only has to provide the coupled_ode_system with analytic gradients and he will speed up stiff and non-stiff integration).

We've never supported that, right?

File: stan/math/rev/mat/functor/cvodes_integrator.hpp [hunk 2]
```diff
@@ -134,12 +112,15 @@ class cvodes_integrator {
     if (cvodes_mem == NULL)
       throw std::runtime_error("CVodeCreate failed to allocate memory");
 
+    const size_t size
```
Comment (syclik) [inline stan/math/rev/mat/functor/cvodes_integrator.hpp]: it looks like this is just used in line 119. Mind either:
- renaming to `coupled_size` and removing the comment
or
- just use it directly in line 119?

**Comment by wds15** [comment]
"@wds15, the first level high level point is... what's being refactored? I think stan/math/rev/mat/functor/cvodes_integrator.hpp's integrate() function keeps its interface and most of its behavior intact. Is that it?"

Yes. Everything stays the same as to how to use that `integrate` function. This function itself has some minor changes, but the biggest impact of this refactor is on `cvodes_ode_data` which is now made to use use `coupled_ode_system` in order to build what CVODES needs as sensitivity RHS of the ODE (recall that CVODES discriminates between the vanilla ODE and the sensitivity ODE RHS). I will answer the rest later.

**Comment by syclik** [comment]
Can you summarize or guide me through the changes made on `cvodes_ode_data`?
Are there no tests for `cvodes_ode_data`?

**Comment by bob-carpenter** [comment]
We've never supported user-defined gradients anywhere in Stan.  I think it's something we should be thinking about for efficiency.  Especially for things like these ODE systems.

**Comment by wds15** [comment]
I enumerated the points above, so in order to your questions @syclik 

1. I am using the one here: `https://github.com/stan-dev/cmdstan/tree/feature/proto-mpi/examples/mpi` Just take out the `map_rect` stuff from it and run it with the `map_rect_stan` function. Of course, I am testing the bdf integrator.

2. OK

3. The `dense_jacobian` function in `cvodes_ode_data`

4. No, we never supported user supplied gradients. It has always been a wish of mine and I have even R code available which code-generates the Jacobians, but ultimatley this approach was never accepted. So I think we should just go with `coupled_ode_system` and if we think we want to supply the gradients, then we just write a version of `coupled_ode_system` which uses internally jacobians. We can worry about that should we really come to that point once; but honestly: with MPI I am just going to throw CPUs at these ODE problems. Problem solved (it's anyway a cold winter after all...).

File: test/unit/math/rev/mat/functor/integrate_ode_adams_prim_test.cpp (hunk 1 above)
Comment (wds15) [reply test/unit/math/rev/mat/functor/integrate_ode_adams_prim_test.cpp]: This error message is aligned with the non-stiff system. I agree that the old error message is easier to understand, but then we would need to change the non-stiff system as well and in addition we cannot use anymore our check functions here. I suggest to either (a) leave it in this pull for now as is and we make another issue+pull for improving error messages in both systems or (b) we need to touch the non-stiff system as well. I would prefer a.

File: stan/math/rev/mat/functor/cvodes_integrator.hpp (hunk 2 above)
Comment (wds15) [reply stan/math/rev/mat/functor/cvodes_integrator.hpp]: Sure, will do.

**Comment by wds15** [comment]
More info is now in the description for the `cvodes_ode_data`; i hope those are sufficiently clear. Let me know if more explanations are needed.

**Comment by bob-carpenter** [comment]
> On Feb 16, 2018, at 3:41 AM, wds15 <notifications@github.com> wrote:
> 
> ...
> 	• No, we never supported user supplied gradients. It has always been a wish of mine and I have even R code available which code-generates the Jacobians, but ultimatley this approach was never accepted.

I want to revisit this.  Not the code-generator thing, which is outside our ability to integrate, but the ability to supply user-generated gradients.  I'm not as pessimistic as Michael, especially if we can add some automatic testing.

> So I think we should just go with coupled_ode_system and if we think we want to supply the gradients, then we just write a version of coupled_ode_system which uses internally jacobians.

Right---that's what I thought we were doing.

...

> throw CPUs at these ODE problems. Problem solved (it's anyway a cold winter after all...).

Those CPUs do generate a lot of heat!

**Comment by syclik** [comment]
@wds15, I need help running your example to verify that there's any sort of difference in speed.

Can you send specific instructions, preferably something that's just in the Math library, not needing to build CmdStan. (I tried running from CmdStan's `feature/proto-mpi`, but couldn't get it to build; it's missing a `make/mpi` file from the math library -- I don't see it on the `feature/concept-mpi` branch in math either.)

**Comment by syclik** [comment]
Also, the data file is missing on the feature/propto-mpi branch

**Comment by syclik** [comment]
@wds15, I was able to create the data file, but wasn't able to get much further.

Is there a way to isolate the `integrate_ode_*` call?

**Comment by wds15** [comment]
I am still confused why this is required that you run this model.

Later I will post a log file of the run which I did.

What error messages are you getting?

Have you disabled the map_rect calls in the Stan file which are MPI specific? The program I quote runs with vanilla Stan only if you get rid of all map_rect functions except the map_rect_stan function.

**Comment by syclik** [comment]
I don’t need to run that model! That’s the one you said is 20% faster!

**Comment by syclik** [comment]
What I'm trying to do:
1. set up an test case in the math library that I can run in `develop` and this branch to determine if there is a difference in speed. And under which instantiations (C++ template instantiations)
2. verify that it's indeed what you think it is. ~20% speed increase isn't something that's trivial, so if there's something that's that much better in this implementation, we should look for other opportunities to apply that fix or avoid that problem
3. stash the example away for future use when we're able to get performance monitoring for the math library

**Comment by wds15** [comment]
I see. Stashing away this example is a very good investment. This example is my go-to example and it stress tests three things: (1) the ODE system or (2) the matrix exponential or (3) the analytic solution of the system which is a ton of log_sum_exp calls. Let put the example in a form so it runs with vanilla Stan.

**Comment by syclik** [comment]
if you can extract the integrate_ode call so we can run it straight from math, that's better.

**Comment by wds15** [comment]
That is not so straightforward. I do first a simulation to generate parameters per patient and then I simulate fake data.

Moreover, the actual model allows you to test more than just the ODE integrator. It would be very restrictive to only use math.

**Comment by wds15** [comment]
I ADDED to this branch an ODE stress test:

`test/unit/math/rev/mat/functor/integrate_ode_bdf_stress_test.cpp`

This one shows an almost 30% speed increase due to the refactor:

```
cvodes refactored:
[ RUN      ] StanOde_tooMuchWork_test.cvodes_coupled_mm
[       OK ] StanOde_tooMuchWork_test.cvodes_coupled_mm (2210 ms)


current develop:
[ RUN      ] StanOde_tooMuchWork_test.cvodes_coupled_mm
[       OK ] StanOde_tooMuchWork_test.cvodes_coupled_mm (3104 ms)
```

You can execute it with the usual runTest.py. I suggest to remove the test before we merge.

**Comment by seantalts** [comment]
Maybe we can turn it into another performance test we run once a day!

**Comment by syclik** [comment]
@wds15, THIS IS AWESOME!

This clearly shows the performance difference. I'm in agreement with
@seantalts. We need to figure out how to turn this into a performance test,
but it doesn't have to be in this pull request.

Before merging, I do want to find out what's causing the speed-up. It's
important. Funny thing is that the stack is actually bigger on this pull
request. I added a print_stack before the grad call. On develop:

STACK, size=8

0  1.00452:-0.128954  1.00452 : -0.128954

1  0.0231317:0.00162412  0.0231317 : 0.00162412

2  4.39586:0.0294678  4.39586 : 0.0294678

3  0.168774:-0.592177  0.168774 : -0.592177

4  79.3959:2  79.3959 : 2

5  0.388722:2  0.388722 : 2

6  79.7347:1  79.7347 : 1

7  0.049972:0  0.049972 : 0

On feature/issue-760-cvodes-refactor:

STACK, size=22

0  1.00452:-0.128954  1.00452 : -0.128954

1  0.0231317:0.00162412  0.0231317 : 0.00162412

2  4.39586:0.0294678  4.39586 : 0.0294678

3  0.168774:-0.592177  0.168774 : -0.592177

4  79.3959:2  79.3959 : 2

5  0.388722:2  0.388722 : 2

6  79.3964:0  79.3964 : 0

7  0.388265:0  0.388265 : 0

8  79.396:0  79.396 : 0

9  0.388677:0  0.388677 : 0

10  79.396:0  79.396 : 0

11  0.388715:0  0.388715 : 0

12  79.6047:0  79.6047 : 0

13  0.179953:0  0.179953 : 0

14  79.7308:0  79.7308 : 0

15  0.0538636:0  0.0538636 : 0

16  79.7347:0  79.7347 : 0

17  0.0499848:0  0.0499848 : 0

18  79.7347:0  79.7347 : 0

19  0.049972:0  0.049972 : 0

20  79.7347:1  79.7347 : 1

21  0.049972:0  0.049972 : 0

**Comment by syclik** [comment]
Oh. That was for the 1000th run.

**Comment by syclik** [comment]
First off, thanks for putting together that test. It's exactly what we needed on this PR to evaluate things.

I spotted a problem. When I run with all doubles, the old code puts 0 vars on the stack (correct behavior). This pull request puts 14 vars on the stack with all doubles (incorrect behavior).
```
STACK, size=14
0  79.3964:0  79.3964 : 0
1  0.388265:0  0.388265 : 0
2  79.396:0  79.396 : 0
3  0.388677:0  0.388677 : 0
4  79.396:0  79.396 : 0
5  0.388715:0  0.388715 : 0
6  79.6047:0  79.6047 : 0
7  0.179953:0  0.179953 : 0
8  79.7308:0  79.7308 : 0
9  0.0538636:0  0.0538636 : 0
10  79.7347:0  79.7347 : 0
11  0.0499848:0  0.0499848 : 0
12  79.7347:0  79.7347 : 0
13  0.049972:0  0.049972 : 0
```

**Comment by wds15** [comment]
Not sure what you mean with "a problem". I am guessing that you run the benchmark with doubles for the initials and doubles for the parameters. Even in that case the AD stack must get used, because we are running a stiff solver such that the Jacobian is needed. So possible the old code had a problem (which explains the slowness potentially).

Or are u saying that you think there is a memory issue?

**Comment by syclik** [comment]
It's a problem because this function leaves things behind when it shouldn't be. I don't think it's unreasonable to think that a function that has a signature that's `double` return and `double` arguments should leave the autodiff stack in the same state as it was going in. I'm going to claim that there's something wrong when that doesn't happen.

I didn't look closely at how things get allocated onto the stack. Can those variables not be included in a nested call?

**Comment by syclik** [comment]
@wds15, the upgrade to CVODES 3.1 caused some conflicts. Sorry about that. Want to fix it?

Given that we've patched a memory leak in the last couple commits, I want to take a really good look at this again (this week). I'm really glad we found that leak.

**Comment by wds15** [comment]
I will look into it later today.

**Comment by wds15** [comment]
ok, merged the conflicts. I think I also caught a glitch from the cvodes 3.1.0 update which missed to set the callback for the Jacobian wrt to states. I have added the callback back in line 143 of cvodes_integrator.h (so it was using finite difference as a default).

@yizhang-cae maybe have a look as well on the line I quote. Thanks.

**Comment by wds15** [comment]
finally all the test systems are happy again. So this is ready for review.

**Comment by wds15** [comment]
Looking at the code again I think we should refactor the cvodes integration in a second step. That is, right now we have a cvodes_ode_data and a cvodes_integrator class. These two should be possible to merge into a single class. However, I suggest to leave the current pull as it stands since this concentrates on merging the stiff and non-stiff ode system.

And after all, we should bring in this pull sooner than later since this gives us some speedup and for two it reenables that the Jacobian wrt to states for the stiff integration is calculated using autodiff instead of finite differences.

Does that make sense?

File: stan/math/rev/mat/functor/cvodes_integrator.hpp [hunk 3]
```diff
@@ -158,26 +133,17 @@ class cvodes_integrator {
       // for the stiff solvers we need to reserve additional
       // memory and provide a Jacobian function call
       // new API since 3.0.0: create matrix object and linear solver object
-      SUNMatrix A{SUNDenseMatrix(N, N)};
-      SUNLinearSolver LS{SUNDenseLinearSolver(cvodes_state, A)};
-      cvodes_check_flag(CVDlsSetLinearSolver(cvodes_mem, LS, A),
-                        "CVDlsSetLinearSolver");
+      cvodes_check_flag(
```
Comment (syclik) [inline stan/math/rev/mat/functor/cvodes_integrator.hpp]: @wds15 or @yizhang-cae, can you take a look at these few lines and compare to `develop`? I'm not an expert in how CVODES should be used, but this looks like it's not using CVODES 3.0.

File: stan/math/rev/mat/functor/cvodes_ode_data.hpp [hunk 4]
```diff
@@ -53,12 +68,37 @@ class cvodes_ode_data {
...
-        param_var_ind_(initial_var::value ? N_ : 0),
-        ode_system_(f, value_of(theta), x, x_int, msgs) {}
+        x_(x),
+        x_int_(x_int),
+        msgs_(msgs),
+        S_((initial_var::value ? N_ : 0) + (param_var::value ? M_ : 0)),
+        coupled_ode_(f, y0, theta, x, x_int, msgs),
+        coupled_state_(coupled_ode_.initial_state()),
+        nv_state_(N_VMake_Serial(N_, &coupled_state_[0])),
+        nv_state_sens_(NULL),
```
Comment (syclik) [inline stan/math/rev/mat/functor/cvodes_ode_data.hpp]: we're now using C++11! Want to change this `NULL` to `nullptr`? (If there's other changes being made on this branch; if not, we can hold off.)

File: stan/math/rev/mat/functor/cvodes_ode_data.hpp [hunk 5]
```diff
@@ -70,15 +110,11 @@ class cvodes_ode_data {
...
-                                    NV_DATA_S(y) + explicit_ode->N_);
-    explicit_ode->rhs_sens(explicit_ode->y0_, explicit_ode->theta_, t, y_vec,
-                           yS, ySdot);
+    explicit_ode->rhs_sens(t, NV_DATA_S(y), yS, ySdot);
     return 0;
   }
 
-  static int dense_jacobian(long int N,  // NOLINT(runtime/int)
-                            realtype t, N_Vector y, N_Vector fy, DlsMat J,
+  static int dense_jacobian(realtype t, N_Vector y, N_Vector fy, SUNMatrix J,
```
Comment (syclik) [inline stan/math/rev/mat/functor/cvodes_ode_data.hpp]: Please add some documentation. At least say this is the implementing the type of function that's `CVDlsJacFn` and this is where the doc can be found: <https://computation.llnl.gov/sites/default/files/public/cv_guide.pdf>

File: stan/math/rev/mat/functor/cvodes_ode_data.hpp [hunk 6]
```diff
@@ -53,12 +68,37 @@ class cvodes_ode_data {
...
+
+  ~cvodes_ode_data() {
+    SUNLinSolFree(LS_);
+    SUNMatDestroy(A_);
+    N_VDestroy_Serial(nv_state_);
+    if (S_ > 0)
+      N_VDestroyVectorArray_Serial(nv_state_sens_, S_);
+  }
 
   static int ode_rhs(double t, N_Vector y, N_Vector ydot, void* user_data) {
```
Comment (syclik) [inline stan/math/rev/mat/functor/cvodes_ode_data.hpp]: Please document this function too. Especially what CVODES function type this is implementing.

File: stan/math/rev/mat/functor/cvodes_ode_data.hpp [hunk 7]
```diff
@@ -88,119 +124,41 @@ class cvodes_ode_data {
  private:
   inline void rhs(const double y[], double dy_dt[], double t) const {
```
Comment (syclik) [inline stan/math/rev/mat/functor/cvodes_ode_data.hpp]: Please document this method. Especially the fact that the output of this function is `dy_dy`. Is there any reason the inputs are not first followed by the output?

File: stan/math/rev/mat/functor/cvodes_ode_data.hpp [hunk 8]
```diff
@@ -88,119 +124,41 @@ class cvodes_ode_data {
...
+        = f_(t, y_vec, theta_dbl_, x_, x_int_, msgs_);
+    check_size_match("cvodes_ode_data", "dz_dt", dy_dt_vec.size(), "states",
+                     N_);
+    std::copy(dy_dt_vec.begin(), dy_dt_vec.end(), dy_dt);
   }
 
-  inline int dense_jacobian(const double* y, DlsMat J, double t) const {
+  // the Jacobian of the ODE system is the coupled ode system for
+  // varying states evaluated at the state y.
+  inline int dense_jacobian(const double* y, SUNMatrix J, double t) const {
```
Comment (syclik) [inline stan/math/rev/mat/functor/cvodes_ode_data.hpp]: Doc please.

File: stan/math/rev/mat/functor/cvodes_ode_data.hpp [hunk 9]
```diff
@@ -88,119 +124,41 @@ class cvodes_ode_data {
...
-   */
-  void rhs_sens(const std::vector<var>& initial,
-                const std::vector<double>& param, double t,
-                const std::vector<double>& y, N_Vector* yS,
-                N_Vector* ySdot) const {
-    Eigen::VectorXd dy_dt(N_);
-    Eigen::MatrixXd Jy(N_, N_);
-    ode_system_.jacobian(t, y, dy_dt, Jy);
-    rhs_sens_initial(Jy, yS, ySdot);
+  inline void rhs_sens(double t, double y[], N_Vector* yS,
```
Comment (syclik) [inline stan/math/rev/mat/functor/cvodes_ode_data.hpp]: doc method, please.

File: test/unit/math/rev/mat/functor/integrate_ode_bdf_stress_test.cpp [hunk 10]
```diff
@@ -0,0 +1,55 @@
+#include <stan/math/rev/mat.hpp>
```
Comment (syclik) [inline test/unit/math/rev/mat/functor/integrate_ode_bdf_stress_test.cpp]: TODO for me: save this test somewhere else and remove it from here.

File: stan/math/rev/mat/functor/cvodes_integrator.hpp (hunk 3 above)
Comment (wds15) [reply stan/math/rev/mat/functor/cvodes_integrator.hpp]: I have moved the resource allocation things into `cvodes_ode_data`. This ensures proper resource deallocation whenever we throw an exception. I think that the current version in develop leaks memory whenever an exception is fired from the ODE RHS since those additional matrices are not taken off the memory. This is now ensured by using RAII and relying essentially on the destructor of `cvodes_ode_data` which will be called even if an exception is fired.

File: stan/math/rev/mat/functor/cvodes_ode_data.hpp (hunk 4 above)
Comment (wds15) [reply stan/math/rev/mat/functor/cvodes_ode_data.hpp]: Changed it.

File: stan/math/rev/mat/functor/cvodes_ode_data.hpp (hunk 5 above)
Comment (wds15) [reply stan/math/rev/mat/functor/cvodes_ode_data.hpp]: I linked the cvs_guide.pdf which also covers the sensitivity RHS.

File: stan/math/rev/mat/functor/cvodes_ode_data.hpp (hunk 6 above)
Comment (wds15) [reply stan/math/rev/mat/functor/cvodes_ode_data.hpp]: I kept it brief. What you asked for is there.

File: stan/math/rev/mat/functor/cvodes_ode_data.hpp (hunk 7 above)
Comment (wds15) [reply stan/math/rev/mat/functor/cvodes_ode_data.hpp]: Doc is there. The order is aligned (along with other functions which were not quite straight).

File: stan/math/rev/mat/functor/cvodes_ode_data.hpp (hunk 8 above)
Comment (wds15) [reply stan/math/rev/mat/functor/cvodes_ode_data.hpp]: Brief doc is in there now.

File: stan/math/rev/mat/functor/cvodes_ode_data.hpp (hunk 9 above)
Comment (wds15) [reply stan/math/rev/mat/functor/cvodes_ode_data.hpp]: OK

**Comment by wds15** [comment]
@seantalts would you mind to check the Jenkins box. I am getting in this and in the mpi cluster pull errors due to distribution tests with the same error message which looks to me like a local problem on the machine. Thanks!

The error message is:
```
remote file operation failed: /Users/Shared/Jenkins/gelman-group-mac/workspace/Math_Pipeline_PR-761-LLPTRJQJGPJJVXWVD6JVO7QLTETYUC2XAFUX7WCPNXHSWTYVCAUA/dist.log.zip at hudson.remoting.Channel@18dfe087:gelman-group-mac: hudson.remoting.ChannelClosedException: Remote call on gelman-group-mac failed. The channel is closing down or has closed down
```

**Comment by seantalts** [comment]
Yeah, I have no idea why those are happening - active area of research. For
now can you retry? It should be intermittent?

**Comment by wds15** [comment]
looks like now its all fine.

File: stan/math/rev/mat/functor/cvodes_ode_data.hpp (hunk 5 above)
Comment (syclik) [reply stan/math/rev/mat/functor/cvodes_ode_data.hpp]: Thanks! That helps a ton. It's still a little short on the standards of our doc, but I think this is enough for now. I wish that doc were there before -- it would have saved me a few hours while reading this code.

**Comment by wds15** [comment]
Great, thanks.

I would then go ahead and merge once lights are green again... unless there is an objection in the meantime.
